protocol_connect.go

```diff
@@ -144,6 +180,38 @@ func (h *connectHandler) NewConn(
 		}
 	}
 
+	var requestBody io.ReadCloser
+	var contentType, codecName string
+	if request.Method == http.MethodGet {
+		if failed == nil && !query.Has(connectUnaryEncodingQueryParameter) {
+			failed = errorf(CodeInvalidArgument, "missing %s parameter", connectUnaryEncodingQueryParameter)
+		} else if failed == nil && !query.Has(connectUnaryMessageQueryParameter) {
+			failed = errorf(CodeInvalidArgument, "missing %s parameter", connectUnaryMessageQueryParameter)
+		}
+		msg := query.Get(connectUnaryMessageQueryParameter)
+		msgReader := queryValueReader(msg, query.Get(connectUnaryBase64QueryParameter) == "1")
+		requestBody = io.NopCloser(msgReader)
+		codecName = query.Get(connectUnaryEncodingQueryParameter)
+		contentType = connectContentTypeFromCodecName(
+			h.Spec.StreamType,
+			codecName,
+		)
+	} else {
+		requestBody = request.Body
+		contentType = getHeaderCanonical(request.Header, headerContentType)
+		codecName = connectCodecFromContentType(
+			h.Spec.StreamType,
+			contentType,
+		)
+	}
+
+	codec := h.Codecs.Get(codecName)
+	// The codec can be nil in the GET request case; that's okay: when failed
+	// is non-nil, codec is never used.
+	if failed == nil && codec == nil {
+		failed = errorf(CodeInvalidArgument, "invalid message encoding: %q", codecName)
+	}
```

Comment:
Acknowledged, but I think maybe this could use more thought: semantically speaking, 415 feels like something that refers to the `POST` body or at least the headers for the `POST` body. Is it semantically correct to return a 415 for this? If you would still prefer that, I'm curious what you think would be a good way to plumb this error through such that we can respond specially for it.

Reply:
I agree that a 415 feels semantically odd - whether it's appropriate depends on how liberally you want to read, ["The format problem might be due to...inspecting the data directly."](https://www.rfc-editor.org/rfc/rfc9110.html#name-415-unsupported-media-type) At best, we're on the border of what's acceptable.

But if we don't return a 415 here, we're now into territory where error handling differs based on protocol: gRPC servers should _always_ return a 415 (which clients should interpret as `CodeInternal`) if they don't understand the serialization scheme, but Connect servers may return a 415 or a 400 (with an explicit `CodeInvalidArgument`) depending on HTTP method. This waters down the "change protocols with a config flag, no other code changes required" story.

I think we may be able to do return 415s by having each protocol implement a `CanHandleGetCodec(query http.Header) bool` method, and have `handler.go` return a 415 if it fails. We could call that method in the switch in `handler.go` that selects the `protocolHandler`.

Reply:
I agree especially to keep semantics the same across GET/POST it would probably be necessary to handle this case the same, otherwise it creates a headache for translation. I gave a shot at it. da71ecab4b5cf49a49ca0ef192f3bac82e80c93b